In [1]:
import pandas as pd

In [8]:
titanic_data = pd.read_csv('Documents/titanic.csv', quotechar='"')
titanic_data.head()

Passengerid   Age     Fare  Sex  sibsp  zero  zero.1  zero.2  zero.3  \
0            1  22.0   7.2500    0      1     0       0       0       0   
1            2  38.0  71.2833    1      1     0       0       0       0   
2            3  26.0   7.9250    1      0     0       0       0       0   
3            4  35.0  53.1000    1      1     0       0       0       0   
4            5  35.0   8.0500    0      0     0       0       0       0   

   zero.4  ...  zero.12  zero.13  zero.14  Pclass  zero.15  zero.16  Embarked  \
0       0  ...        0        0        0       3        0        0       2.0   
1       0  ...        0        0        0       1        0        0       0.0   
2       0  ...        0        0        0       3        0        0       2.0   
3       0  ...        0        0        0       1        0        0       2.0   
4       0  ...        0        0        0       3        0        0       2.0   

   zero.17  zero.18  2urvived  
0        0        0         0  
1        0        0         1  
2        0        0         1  
3        0        0         1  
4        0        0         0  

[5 rows x 28 columns]

In [11]:
columns_to_drop = ['Passengerid', 'zero', 'zero.1', 'zero.1', 'zero.2', 'zero.3', 'zero.4', 'zero.5', 'zero.6', 'zero.7', 'zero.8', 'zero.9', 'zero.10', 'zero.11', 'zero.12', 'zero.13', 'zero.14', 'zero.15', 'zero.16', 'zero.17', 'zero.18']
titanic_data.drop(columns=columns_to_drop, inplace=True)
titanic_data.head()


Age     Fare  Sex  sibsp  Parch  Pclass  Embarked  2urvived
0  22.0   7.2500    0      1      0       3       2.0         0
1  38.0  71.2833    1      1      0       1       0.0         1
2  26.0   7.9250    1      0      0       3       2.0         1
3  35.0  53.1000    1      1      0       1       2.0         1
4  35.0   8.0500    0      0      0       3       2.0         0

In [12]:
titanic_data = pd.get_dummies(titanic_data, columns=['Embarked'])
titanic_data.head()

Age     Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
0  22.0   7.2500    0      1      0       3         0         False   
1  38.0  71.2833    1      1      0       1         1          True   
2  26.0   7.9250    1      0      0       3         1         False   
3  35.0  53.1000    1      1      0       1         1         False   
4  35.0   8.0500    0      0      0       3         0         False   

   Embarked_1.0  Embarked_2.0  
0         False          True  
1         False         False  
2         False          True  
3         False          True  
4         False          True

In [13]:
titanic_data.replace({True: 1, False: 0}, inplace=True)


In [14]:
titanic_data.head()

Age     Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
0  22.0   7.2500    0      1      0       3         0             0   
1  38.0  71.2833    1      1      0       1         1             1   
2  26.0   7.9250    1      0      0       3         1             0   
3  35.0  53.1000    1      1      0       1         1             0   
4  35.0   8.0500    0      0      0       3         0             0   

   Embarked_1.0  Embarked_2.0  
0             0             1  
1             0             0  
2             0             1  
3             0             1  
4             0             1

Now l'ets see if is there any invalid values in our dataset.

In [19]:
titanic_data.head()

Age     Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
0  22.0   7.2500    0      1      0       3         0             0   
1  38.0  71.2833    1      1      0       1         1             1   
2  26.0   7.9250    1      0      0       3         1             0   
3  35.0  53.1000    1      1      0       1         1             0   
4  35.0   8.0500    0      0      0       3         0             0   

   Embarked_1.0  Embarked_2.0  
0             0             1  
1             0             0  
2             0             1  
3             0             1  
4             0             1

In [20]:
titanic_data[titanic_data.isnull().any(axis=1)]

Empty DataFrame
Columns: [Age, Fare, Sex, sibsp, Parch, Pclass, 2urvived, Embarked_0.0, Embarked_1.0, Embarked_2.0]
Index: []

In [26]:
from IPython.display import display

# Assuming titanic_data is your DataFrame
invalid_rows = titanic_data[titanic_data.isnull().any(axis=1)]

# Sample 10 rows for display
sampled_rows = invalid_rows.sample(min(10, len(invalid_rows)))

# Display the sampled rows
print(sampled_rows)

Empty DataFrame
Columns: [Age, Fare, Sex, sibsp, Parch, Pclass, 2urvived, Embarked_0.0, Embarked_1.0, Embarked_2.0]
Index: []


In [27]:
nan_values = titanic_data[titanic_data.isnull().any(axis=1)]
print(nan_values)

Empty DataFrame
Columns: [Age, Fare, Sex, sibsp, Parch, Pclass, 2urvived, Embarked_0.0, Embarked_1.0, Embarked_2.0]
Index: []


In [28]:
from sklearn.cluster import MeanShift

analyzer = MeanShift(bandwidth=50)
analyzer.fit(titanic_data)

MeanShift(bandwidth=50)

there is a function that estimates the bandwidth :

In [29]:
from sklearn.cluster import estimate_bandwidth
estimate_bandwidth(titanic_data)

28.231640426987656

In [30]:
import numpy as np

Now we see how many cluster we have :

In [34]:
labels = analyzer.labels_

In [35]:
import numpy as np
np.unique(labels)

array([0, 1, 2], dtype=int64)

Now we will add a new column called cluster group to our data frame. This column will contain information as to what cluster a particular passenger belongs to.

In [37]:
import numpy as np

titanic_data['cluster_group'] = np.nan
data_length = len(titanic_data)
for i in range(data_length):
    titanic_data.iloc[i, titanic_data.columns.get_loc('cluster_group')] = labels[i]

In [38]:
titanic_data.head()

Age     Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
0  22.0   7.2500    0      1      0       3         0             0   
1  38.0  71.2833    1      1      0       1         1             1   
2  26.0   7.9250    1      0      0       3         1             0   
3  35.0  53.1000    1      1      0       1         1             0   
4  35.0   8.0500    0      0      0       3         0             0   

   Embarked_1.0  Embarked_2.0  cluster_group  
0             0             1            0.0  
1             0             0            0.0  
2             0             1            0.0  
3             0             1            0.0  
4             0             1            0.0

In [39]:
titanic_data.describe()

Age         Fare          Sex        sibsp        Parch  \
count  1309.000000  1309.000000  1309.000000  1309.000000  1309.000000   
mean     29.503186    33.281086     0.355997     0.498854     0.385027   
std      12.905241    51.741500     0.478997     1.041658     0.865560   
min       0.170000     0.000000     0.000000     0.000000     0.000000   
25%      22.000000     7.895800     0.000000     0.000000     0.000000   
50%      28.000000    14.454200     0.000000     0.000000     0.000000   
75%      35.000000    31.275000     1.000000     1.000000     0.000000   
max      80.000000   512.329200     1.000000     8.000000     9.000000   

            Pclass     2urvived  Embarked_0.0  Embarked_1.0  Embarked_2.0  \
count  1309.000000  1309.000000   1309.000000   1309.000000   1309.000000   
mean      2.294882     0.261268      0.206264      0.093965      0.698243   
std       0.837836     0.439494      0.404777      0.291891      0.459196   
min       1.000000     0.000000      0.000000      0.000000      0.000000   
25%       2.000000     0.000000      0.000000      0.000000      0.000000   
50%       3.000000     0.000000      0.000000      0.000000      1.000000   
75%       3.000000     1.000000      0.000000      0.000000      1.000000   
max       3.000000     1.000000      1.000000      1.000000      1.000000   

       cluster_group  
count    1309.000000  
mean        0.054240  
std         0.239694  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         2.000000

Now we will regroupe the clusters by using the cluster group column that we just added and then find the mean for each cluster.
This will be the average of all characteristics within each group.

In [40]:
titanic_cluster_data = titanic_data.groupby(['cluster_group']).mean()
titanic_cluster_data

Age        Fare       Sex     sibsp     Parch    Pclass  \
cluster_group                                                                  
0.0            29.160427   23.515489  0.340580  0.490338  0.359903  2.364734   
1.0            35.530476  195.387103  0.650794  0.698413  0.873016  1.000000   
2.0            41.000000  512.329200  0.500000  0.000000  0.500000  1.000000   

               2urvived  Embarked_0.0  Embarked_1.0  Embarked_2.0  
cluster_group                                                      
0.0            0.252818      0.187601      0.099034      0.711755  
1.0            0.396825      0.523810      0.000000      0.476190  
2.0            0.750000      1.000000      0.000000      0.000000

The average informations will be much valuable if we add the number of samples that exist in each cluster.
Now we will add a counts column to our data frame and set its value to the size of each cluster.

In [41]:
titanic_cluster_data['Counts'] = pd.Series(titanic_data.groupby(['cluster_group']).size())
titanic_cluster_data

Age        Fare       Sex     sibsp     Parch    Pclass  \
cluster_group                                                                  
0.0            29.160427   23.515489  0.340580  0.490338  0.359903  2.364734   
1.0            35.530476  195.387103  0.650794  0.698413  0.873016  1.000000   
2.0            41.000000  512.329200  0.500000  0.000000  0.500000  1.000000   

               2urvived  Embarked_0.0  Embarked_1.0  Embarked_2.0  Counts  
cluster_group                                                              
0.0            0.252818      0.187601      0.099034      0.711755    1242  
1.0            0.396825      0.523810      0.000000      0.476190      63  
2.0            0.750000      1.000000      0.000000      0.000000       4

The first group 0.0 has 1242 who did not survive (survival rate is around 25%), they belong to the lower classes, second and third classes mostly
The sex column shows that there as more males than females in this cluster 65%. The average fare paid by the passengers is around 23£.


    The second cluster comprises mostly of people who survived the Titanic disaster. their survival rate is around 40%. We notice that they are from first class and mostly females 65%. Also, the average fare paid by these passengers is 195£ far higher than the previous cluster.

However, the last cluster is not really significant because it has a very small number,thus we cant work on it.

In [45]:
titanic_data[titanic_data['cluster_group'] == 1 ]

Age      Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
27    19.0  263.0000    0      3      2       1         0             0   
31    28.0  146.5208    1      1      0       1         1             1   
88    23.0  263.0000    1      3      2       1         1             0   
118   24.0  247.5208    0      0      1       1         0             1   
195   58.0  146.5208    1      0      0       1         1             1   
...    ...       ...  ...    ...    ...     ...       ...           ...   
1215  39.0  211.3375    1      0      0       1         0             0   
1262  31.0  134.5000    1      0      0       1         0             1   
1266  45.0  262.3750    1      0      0       1         0             1   
1291  30.0  164.8667    1      0      0       1         0             0   
1298  50.0  211.5000    0      1      1       1         0             1   

      Embarked_1.0  Embarked_2.0  cluster_group  
27               0             1            1.0  
31               0             0            1.0  
88               0             1            1.0  
118              0             0            1.0  
195              0             0            1.0  
...            ...           ...            ...  
1215             0             1            1.0  
1262             0             0            1.0  
1266             0             0            1.0  
1291             0             1            1.0  
1298             0             0            1.0  

[63 rows x 11 columns]

We will filter the contents of the data frame to see all the passengers that belong to this cluster.

Now we will change the bandwidth to be 30 

In [49]:
from sklearn.cluster import MeanShift

analyzer = MeanShift(bandwidth=30)
analyzer.fit(titanic_data)

MeanShift(bandwidth=30)

In [50]:
from sklearn.cluster import estimate_bandwidth
estimate_bandwidth(titanic_data)

28.231812487379027

In [51]:
import numpy as np
labels = analyzer.labels_

In [52]:
import numpy as np
np.unique(labels)

array([0, 1, 2, 3, 4], dtype=int64)

In [53]:
import numpy as np

titanic_data['cluster_group'] = np.nan
data_length = len(titanic_data)
for i in range(data_length):
    titanic_data.iloc[i, titanic_data.columns.get_loc('cluster_group')] = labels[i]

In [54]:
titanic_data.head()

Age     Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
0  22.0   7.2500    0      1      0       3         0             0   
1  38.0  71.2833    1      1      0       1         1             1   
2  26.0   7.9250    1      0      0       3         1             0   
3  35.0  53.1000    1      1      0       1         1             0   
4  35.0   8.0500    0      0      0       3         0             0   

   Embarked_1.0  Embarked_2.0  cluster_group  
0             0             1            0.0  
1             0             0            1.0  
2             0             1            0.0  
3             0             1            0.0  
4             0             1            0.0

In [55]:
titanic_cluster_data = titanic_data.groupby(['cluster_group']).mean()
titanic_cluster_data

Age        Fare       Sex     sibsp     Parch    Pclass  \
cluster_group                                                                  
0.0            28.550087   18.624616  0.325239  0.478751  0.355594  2.464007   
1.0            36.406102  101.327647  0.567797  0.610169  0.466102  1.059322   
2.0            37.805556  218.199544  0.611111  0.333333  0.333333  1.000000   
3.0            35.062500  259.824212  0.687500  1.437500  1.937500  1.000000   
4.0            41.000000  512.329200  0.500000  0.000000  0.500000  1.000000   

               2urvived  Embarked_0.0  Embarked_1.0  Embarked_2.0  
cluster_group                                                      
0.0            0.230703      0.163920      0.104076      0.732003  
1.0            0.525424      0.483051      0.025424      0.474576  
2.0            0.333333      0.555556      0.000000      0.444444  
3.0            0.312500      0.625000      0.000000      0.375000  
4.0            0.750000      1.000000      0.000000      0.000000

In [56]:
titanic_cluster_data['Counts'] = pd.Series(titanic_data.groupby(['cluster_group']).size())
titanic_cluster_data

Age        Fare       Sex     sibsp     Parch    Pclass  \
cluster_group                                                                  
0.0            28.550087   18.624616  0.325239  0.478751  0.355594  2.464007   
1.0            36.406102  101.327647  0.567797  0.610169  0.466102  1.059322   
2.0            37.805556  218.199544  0.611111  0.333333  0.333333  1.000000   
3.0            35.062500  259.824212  0.687500  1.437500  1.937500  1.000000   
4.0            41.000000  512.329200  0.500000  0.000000  0.500000  1.000000   

               2urvived  Embarked_0.0  Embarked_1.0  Embarked_2.0  Counts  
cluster_group                                                              
0.0            0.230703      0.163920      0.104076      0.732003    1153  
1.0            0.525424      0.483051      0.025424      0.474576     118  
2.0            0.333333      0.555556      0.000000      0.444444      18  
3.0            0.312500      0.625000      0.000000      0.375000      16  
4.0            0.750000      1.000000      0.000000      0.000000       4

In [57]:
titanic_data[titanic_data['cluster_group'] == 1 ]

Age      Fare  Sex  sibsp  Parch  Pclass  2urvived  Embarked_0.0  \
1     38.0   71.2833    1      1      0       1         1             1   
31    28.0  146.5208    1      1      0       1         1             1   
34    28.0   82.1708    0      1      0       1         0             1   
52    49.0   76.7292    1      1      0       1         1             1   
61    38.0   80.0000    1      0      0       1         1             0   
...    ...       ...  ...    ...    ...     ...       ...           ...   
1281  23.0   93.5000    0      0      0       1         0             0   
1288  48.0   79.2000    1      1      1       1         0             1   
1291  30.0  164.8667    1      0      0       1         0             0   
1302  37.0   90.0000    1      1      0       1         0             0   
1305  39.0  108.9000    1      0      0       1         0             1   

      Embarked_1.0  Embarked_2.0  cluster_group  
1                0             0            1.0  
31               0             0            1.0  
34               0             0            1.0  
52               0             0            1.0  
61               0             0            1.0  
...            ...           ...            ...  
1281             0             1            1.0  
1288             0             0            1.0  
1291             0             1            1.0  
1302             1             0            1.0  
1305             0             0            1.0  

[118 rows x 11 columns]